File: Harvest/Services/CompatibilityService.swift
import Foundation

struct CompatibilityService {
    /// Calculate compatibility score between two users
    /// Returns a score from 0-100 based on multiple factors
    func calculateCompatibility(
        currentUser: UserProfile,
        otherUser: UserProfile,
        currentUserAxisScores: (need: AxisScores, bring: AxisScores)? = nil,
        otherUserAxisScores: (need: AxisScores, bring: AxisScores)? = nil
    ) -> CompatibilityScore {
        var totalScore = 0.0
        var breakdown: [String: Double] = [:]

        let interestsScore = calculateInterestsScore(
            userHobbies: currentUser.hobbies ?? [],
            otherHobbies: otherUser.hobbies ?? []
        )
        breakdown["interests"] = interestsScore
        totalScore += interestsScore

        let valuesScore = calculateValuesScore(
            currentUserAxisScores: currentUserAxisScores,
            otherUserAxisScores: otherUserAxisScores
        )
        breakdown["values"] = valuesScore
        totalScore += valuesScore

        let goalsScore = calculateGoalsScore(
            userGoals: currentUser.goalsList,
            otherGoals: otherUser.goalsList
        )
        breakdown["goals"] = goalsScore
        totalScore += goalsScore

        let ageScore: Double
        if let userAge = currentUser.age, let otherAge = otherUser.age {
            ageScore = calculateAgeScore(userAge: userAge, otherAge: otherAge)
        } else {
            ageScore = 5.0
        }
        breakdown["age"] = ageScore
        totalScore += ageScore

        breakdown["distance"] = 5.0
        totalScore += 5.0

        return CompatibilityScore(
            total: Int(min(100, totalScore)),
            breakdown: breakdown
        )
    }

    // MARK: - Interest/Hobby Scoring

    private func calculateInterestsScore(userHobbies: [String], otherHobbies: [String]) -> Double {
        guard !userHobbies.isEmpty, !otherHobbies.isEmpty else {
            return 20.0 // Neutral score if no hobbies listed
        }

        let userSet = Set(userHobbies.map { $0.lowercased() })
        let otherSet = Set(otherHobbies.map { $0.lowercased() })
        let sharedCount = userSet.intersection(otherSet).count

        // Calculate Jaccard similarity
        let unionCount = userSet.union(otherSet).count
        let similarity = Double(sharedCount) / Double(unionCount)

        // Scale to 0-40 points
        // 0% shared = 0 points
        // 25% shared = 20 points
        // 50%+ shared = 40 points
        let score = min(40.0, similarity * 80.0)

        return score
    }

    // MARK: - Values Scoring

    private func calculateValuesScore(
        currentUserAxisScores: (need: AxisScores, bring: AxisScores)?,
        otherUserAxisScores: (need: AxisScores, bring: AxisScores)?
    ) -> Double {
        guard let me = currentUserAxisScores,
              let them = otherUserAxisScores,
              !me.need.isZero, !me.bring.isZero,
              !them.need.isZero, !them.bring.isZero else {
            return 15.0
        }
        let avgCosine =
            (AxisScores.cosine(me.need, them.bring) + AxisScores.cosine(me.bring, them.need)) / 2.0
        // Cosine on non-negative vectors is in [0, 1] -> scale to 0...30
        return max(0, avgCosine) * 30.0
    }

    // MARK: - Goals Scoring

    private func calculateGoalsScore(userGoals: [String], otherGoals: [String]) -> Double {
        guard !userGoals.isEmpty, !otherGoals.isEmpty else {
            return 7.5 // Neutral score
        }

        let userSet = Set(userGoals.map { $0.lowercased() })
        let otherSet = Set(otherGoals.map { $0.lowercased() })
        let sharedCount = userSet.intersection(otherSet).count

        // Having even one shared goal is significant
        // 0 shared = 0 points
        // 1 shared = 10 points
        // 2+ shared = 15 points
        if sharedCount == 0 {
            return 0
        } else if sharedCount == 1 {
            return 10.0
        } else {
            return 15.0
        }
    }

    // MARK: - Age Scoring

    private func calculateAgeScore(userAge: Int, otherAge: Int) -> Double {
        let ageDifference = abs(userAge - otherAge)

        // Age difference scoring:
        // 0-2 years: 10 points
        // 3-5 years: 7 points
        // 6-10 years: 4 points
        // 11+ years: 2 points
        switch ageDifference {
        case 0...2:
            return 10.0
        case 3...5:
            return 7.0
        case 6...10:
            return 4.0
        default:
            return 2.0
        }
    }

    // MARK: - Ranking Profiles

    /// Sort profiles by compatibility score (highest first)
    func rankProfiles(
        currentUser: UserProfile,
        profiles: [UserProfile],
        currentUserAxisScores: (need: AxisScores, bring: AxisScores)?,
        otherUsersAxisScores: [String: (need: AxisScores, bring: AxisScores)]
    ) -> [(profile: UserProfile, score: CompatibilityScore)] {
        var scoredProfiles: [(profile: UserProfile, score: CompatibilityScore)] = []

        for profile in profiles {
            let otherAxis = otherUsersAxisScores[profile.id]
            let score = calculateCompatibility(
                currentUser: currentUser,
                otherUser: profile,
                currentUserAxisScores: currentUserAxisScores,
                otherUserAxisScores: otherAxis
            )
            scoredProfiles.append((profile: profile, score: score))
        }

        return scoredProfiles.sorted { $0.score.total > $1.score.total }
    }

    // MARK: - Value-pick overlap

    /// Intersection of value picks between two users on each side.
    /// `theyBringForMyNeeds`: which of their Bring picks satisfy my Need picks.
    /// `iBringForTheirNeeds`: which of my Bring picks satisfy their Need picks.
    struct ValueOverlap: Sendable {
        let theyBringForMyNeeds: [Value]
        let iBringForTheirNeeds: [Value]
    }

    func valueOverlap(
        myNeeds: [Value],
        myBrings: [Value],
        theirNeeds: [Value],
        theirBrings: [Value]
    ) -> ValueOverlap {
        let myNeedIds = Set(myNeeds.map(\.id))
        let theirNeedIds = Set(theirNeeds.map(\.id))
        let theyForMe = theirBrings.filter { myNeedIds.contains($0.id) }
        let meForThem = myBrings.filter { theirNeedIds.contains($0.id) }
        return ValueOverlap(
            theyBringForMyNeeds: theyForMe,
            iBringForTheirNeeds: meForThem
        )
    }

    // MARK: - Compatibility blurb

    /// One- or two-sentence blurb summarizing how two users align on the 5 axes
    /// plus how their selected value picks line up. Template-driven, no LLM.
    func compatibilityBlurb(
        otherName: String,
        bringCosine: Double,
        needCosine: Double,
        topSharedAxis: ValueAxis?,
        overlap: ValueOverlap,
        myNeedsCount: Int
    ) -> String {
        let strongAlignment = (bringCosine + needCosine) / 2.0 >= 0.6
        let theyForMe = overlap.theyBringForMyNeeds.count

        if let axis = topSharedAxis, strongAlignment, theyForMe > 0 {
            return "You and \(otherName) share a strong foundation around \(axis.displayName) — and what they bring lines up with \(theyForMe) of your \(myNeedsCount) needs."
        }
        if let axis = topSharedAxis, strongAlignment {
            return "You and \(otherName) share a strong foundation around \(axis.displayName)."
        }
        if theyForMe > 0 {
            return "\(otherName) brings \(theyForMe) of your \(myNeedsCount) selected needs."
        }
        return "Your value patterns differ — sometimes that's where growth starts."
    }

    /// Returns the axis where the lower of (myBring[i], theirNeed[i]) is highest,
    /// representing the strongest mutual alignment between bring and need.
    func topSharedAxis(
        myBring: AxisScores,
        theirNeed: AxisScores
    ) -> ValueAxis? {
        let pairs: [(ValueAxis, Double)] = ValueAxis.allCases.map { axis in
            let mine = myBring.value(for: axis)
            let theirs = theirNeed.value(for: axis)
            return (axis, min(mine, theirs))
        }
        let top = pairs.max { $0.1 < $1.1 }
        guard let top, top.1 > 0 else { return nil }
        return top.0
    }
}

// MARK: - Models

struct CompatibilityScore: Sendable {
    let total: Int
    let breakdown: [String: Double]

    var interestsScore: Int {
        Int(breakdown["interests"] ?? 0)
    }

    var valuesScore: Int {
        Int(breakdown["values"] ?? 0)
    }

    var goalsScore: Int {
        Int(breakdown["goals"] ?? 0)
    }

    var ageScore: Int {
        Int(breakdown["age"] ?? 0)
    }

    var distanceScore: Int {
        Int(breakdown["distance"] ?? 0)
    }

    var displayPercentage: String {
        "\(total)%"
    }

    var compatibilityLevel: String {
        switch total {
        case 80...100:
            return "Excellent Match"
        case 60...79:
            return "Great Match"
        case 40...59:
            return "Good Match"
        case 20...39:
            return "Fair Match"
        default:
            return "Low Match"
        }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/data/service/CompatibilityService.kt
package com.harvestglass.harvest.data.service

import com.harvestglass.harvest.data.model.AxisScores
import com.harvestglass.harvest.data.model.UserProfile
import com.harvestglass.harvest.data.model.Value
import com.harvestglass.harvest.data.model.ValueAxis
import kotlin.math.abs
import kotlin.math.min

/** Mirrors CompatibilityScore in Harvest/Services/CompatibilityService.swift. */
data class CompatibilityScore(val total: Int, val breakdown: Map<String, Double>) {
    val interestsScore: Int get() = (breakdown["interests"] ?: 0.0).toInt()
    val valuesScore: Int get() = (breakdown["values"] ?: 0.0).toInt()
    val goalsScore: Int get() = (breakdown["goals"] ?: 0.0).toInt()
    val ageScore: Int get() = (breakdown["age"] ?: 0.0).toInt()
    val distanceScore: Int get() = (breakdown["distance"] ?: 0.0).toInt()

    val displayPercentage: String get() = "$total%"

    val compatibilityLevel: String
        get() = when (total) {
            in 80..100 -> "Excellent Match"
            in 60..79 -> "Great Match"
            in 40..59 -> "Good Match"
            in 20..39 -> "Fair Match"
            else -> "Low Match"
        }
}

/**
 * Which of their Bring picks satisfy my Need picks, and vice versa.
 * Mirrors ValueOverlap in CompatibilityService.swift.
 */
data class ValueOverlap(
    val theyBringForMyNeeds: List<Value>,
    val iBringForTheirNeeds: List<Value>
)

/**
 * Port of Harvest/Services/CompatibilityService.swift.
 *
 * Pure scoring — no network, no state. Every neutral fallback below is the
 * Swift value verbatim; changing one would silently reshuffle everyone's
 * scores relative to iOS.
 */
object CompatibilityService {

    /**
     * A 0–100 score across interests, values, goals, age and distance.
     *
     * Axis scores are optional because a user who hasn't answered the question
     * set has none; that path scores the neutral 15 rather than zero.
     */
    fun calculateCompatibility(
        currentUser: UserProfile,
        otherUser: UserProfile,
        currentUserAxisScores: Pair<AxisScores, AxisScores>? = null,
        otherUserAxisScores: Pair<AxisScores, AxisScores>? = null
    ): CompatibilityScore {
        val breakdown = mutableMapOf<String, Double>()
        var total = 0.0

        val interests = interestsScore(
            currentUser.hobbies.orEmpty(),
            otherUser.hobbies.orEmpty()
        )
        breakdown["interests"] = interests
        total += interests

        val values = valuesScore(currentUserAxisScores, otherUserAxisScores)
        breakdown["values"] = values
        total += values

        val goals = goalsScore(currentUser.goalsList, otherUser.goalsList)
        breakdown["goals"] = goals
        total += goals

        val age = if (currentUser.age != null && otherUser.age != null) {
            ageScore(currentUser.age, otherUser.age)
        } else {
            NEUTRAL_AGE
        }
        breakdown["age"] = age
        total += age

        // Distance is a flat allowance: the app has never scored it.
        breakdown["distance"] = FLAT_DISTANCE
        total += FLAT_DISTANCE

        return CompatibilityScore(total = min(100.0, total).toInt(), breakdown = breakdown)
    }

    /** Jaccard similarity over hobbies, scaled to 0–40. */
    private fun interestsScore(userHobbies: List<String>, otherHobbies: List<String>): Double {
        if (userHobbies.isEmpty() || otherHobbies.isEmpty()) return NEUTRAL_INTERESTS

        val mine = userHobbies.map { it.lowercase() }.toSet()
        val theirs = otherHobbies.map { it.lowercase() }.toSet()
        val union = mine.union(theirs).size
        if (union == 0) return NEUTRAL_INTERESTS

        val similarity = mine.intersect(theirs).size.toDouble() / union
        return min(40.0, similarity * 80.0)
    }

    /**
     * How well my Bring pattern answers their Need pattern and vice versa,
     * averaged and scaled to 0–30.
     */
    private fun valuesScore(
        mine: Pair<AxisScores, AxisScores>?,
        theirs: Pair<AxisScores, AxisScores>?
    ): Double {
        if (mine == null || theirs == null) return NEUTRAL_VALUES
        val (myNeed, myBring) = mine
        val (theirNeed, theirBring) = theirs
        if (myNeed.isZero || myBring.isZero || theirNeed.isZero || theirBring.isZero) {
            return NEUTRAL_VALUES
        }

        val avgCosine =
            (AxisScores.cosine(myNeed, theirBring) + AxisScores.cosine(myBring, theirNeed)) / 2.0
        // Cosine on non-negative vectors is in [0, 1] -> scale to 0...30.
        return maxOf(0.0, avgCosine) * 30.0
    }

    /** Even one shared goal is significant, so this is a step function. */
    private fun goalsScore(userGoals: List<String>, otherGoals: List<String>): Double {
        if (userGoals.isEmpty() || otherGoals.isEmpty()) return NEUTRAL_GOALS

        val shared = userGoals.map { it.lowercase() }.toSet()
            .intersect(otherGoals.map { it.lowercase() }.toSet())
            .size

        return when (shared) {
            0 -> 0.0
            1 -> 10.0
            else -> 15.0
        }
    }

    private fun ageScore(userAge: Int, otherAge: Int): Double =
        when (abs(userAge - otherAge)) {
            in 0..2 -> 10.0
            in 3..5 -> 7.0
            in 6..10 -> 4.0
            else -> 2.0
        }

    /** Profiles sorted by score, highest first. */
    fun rankProfiles(
        currentUser: UserProfile,
        profiles: List<UserProfile>
    ): List<Pair<UserProfile, CompatibilityScore>> =
        profiles
            .map { it to calculateCompatibility(currentUser, it) }
            .sortedByDescending { it.second.total }

    fun valueOverlap(
        myNeeds: List<Value>,
        myBrings: List<Value>,
        theirNeeds: List<Value>,
        theirBrings: List<Value>
    ): ValueOverlap {
        val myNeedIds = myNeeds.map { it.id }.toSet()
        val theirNeedIds = theirNeeds.map { it.id }.toSet()
        return ValueOverlap(
            theyBringForMyNeeds = theirBrings.filter { myNeedIds.contains(it.id) },
            iBringForTheirNeeds = myBrings.filter { theirNeedIds.contains(it.id) }
        )
    }

    /**
     * The axis where the weaker of (my bring, their need) is strongest — the
     * strongest mutual alignment rather than the loudest single voice.
     */
    fun topSharedAxis(myBring: AxisScores, theirNeed: AxisScores): ValueAxis? {
        val top = ValueAxis.entries
            .map { axis -> axis to min(myBring.value(axis), theirNeed.value(axis)) }
            .maxByOrNull { it.second }
            ?: return null

        return if (top.second > 0) top.first else null
    }

    /**
     * One- or two-sentence summary of how two people line up. Template-driven,
     * no model call — same sentences as iOS, chosen in the same order.
     */
    fun compatibilityBlurb(
        otherName: String,
        bringCosine: Double,
        needCosine: Double,
        topSharedAxis: ValueAxis?,
        overlap: ValueOverlap,
        myNeedsCount: Int
    ): String {
        val strongAlignment = (bringCosine + needCosine) / 2.0 >= STRONG_ALIGNMENT
        val theyForMe = overlap.theyBringForMyNeeds.size

        if (topSharedAxis != null && strongAlignment && theyForMe > 0) {
            return "You and $otherName share a strong foundation around " +
                "${topSharedAxis.displayName} — and what they bring lines up with " +
                "$theyForMe of your $myNeedsCount needs."
        }
        if (topSharedAxis != null && strongAlignment) {
            return "You and $otherName share a strong foundation around " +
                "${topSharedAxis.displayName}."
        }
        if (theyForMe > 0) {
            return "$otherName brings $theyForMe of your $myNeedsCount selected needs."
        }
        return "Your value patterns differ — sometimes that's where growth starts."
    }

    // Neutral fallbacks, verbatim from the Swift source.
    private const val NEUTRAL_INTERESTS = 20.0
    private const val NEUTRAL_VALUES = 15.0
    private const val NEUTRAL_GOALS = 7.5
    private const val NEUTRAL_AGE = 5.0
    private const val FLAT_DISTANCE = 5.0
    private const val STRONG_ALIGNMENT = 0.6
}
